arctic/async/async_arctic.py

```diff
@@ -0,0 +1,187 @@
+
+                if is_modifier and self._get_accessors(lib_name, symbol):
+                    raise AsyncArcticException("Can't submit async {} task as one or more {} tasks "
+                                               "are being processed".format(AsyncRequestType.ACCESSOR,
+                                                                            AsyncRequestType.MODIFIER))
+
+            # Create the request object
+            request = AsyncRequest(kind, lib_name, fun, *args, **kwargs)
+
+            # Update the state of tracked tasks
+            self._add_request(request)
+
+            # Submit the task
+            try:
+                new_id, new_future = self.submit_task(False, _arctic_task_exec, request)
+                request.id = new_id
+                request.future = new_future
+            except Exception:
+                # clean up the state
+                self._remove_request(request)
+                raise
+
+        # No need to hold the lock for the below statements
+        # If request is already finished by now, the callback is invoked immediately
+        request.future.add_done_callback(lambda the_future: self._request_finished(request, callback))
+
+        return request
+
+    def _request_finished(self, request, callback=None):
+        with type(self)._POOL_LOCK:
+            self._remove_request(request)
+        request.is_completed = True
+        if callback:
+            callback(request)
+
+    @staticmethod
+    def wait_request(request, timeout=None):
+        while request is not None and not request.is_completed:
+            AsyncArctic.wait_tasks_or_abort((request.future, ), timeout=timeout)
+
+    @staticmethod
+    def wait_requests(requests, timeout=None):
+        while requests and not all(r.is_completed for r in requests):
+            AsyncArctic.wait_tasks_or_abort(tuple(r.future for r in requests if not r.is_completed), timeout=timeout)
+
+    @staticmethod
+    def wait_any_request(requests, timeout=None):
+        while requests and not any(r.is_completed for r in requests):
+            AsyncArctic.wait_tasks(tuple(r.future for r in requests if not r.is_completed),
+                                   timeout=timeout, return_when=FIRST_COMPLETED, raise_exceptions=True)
+
+    @staticmethod
+    def filter_finished_requests(requests, do_raise=True):
+        if not requests:
+            return requests, requests
+        alive_requests = [r for r in requests if not r.is_completed]
+        done_requests = [r for r in requests if r.is_completed]
+        if do_raise:
+            AsyncArctic.raise_errored(done_requests)
+        return alive_requests, done_requests
+
+    @staticmethod
+    def raise_errored(requests):
+        errored = tuple(r for r in requests if r.is_completed and r.exception is not None)
+        if errored:
+            raise errored[0].exception
```

Comment:
Raising just one of the requests might hide the real cause, e.g. a first failing async request may leave some internal functionality in an inconsistent state, but may not be the exception that is finally raised.  Java has suppressed exceptions for exactly this reason, we should do something similar here to allow them all to be introspected.

Reply:
this is a fair point

Internally, the executor will indeed not fail/raise, but rather record the exception:
https://github.com/dimosped/arctic/blob/public-async-final/arctic/async/async_arctic.py#L25

"raise_errored(requests)" is only a convenience static method, not affecting functionally the executor/async_arctic. So it is up to the user to use it to raise any of the requests he provides in the list, it is to quickly raise from a large set of requests which have completed, i.e. raise if any had an exception. This is a separate call, e.g. 

```
requests = []
for s in syms:
    requests.append(async_arctic_submit(lib, lib.write, True, symbol=s, data=data)

async_wait_requests(requests)

try:
    raise_errored(requests)
except:
    # re-run or handle errored ones
    # errored = filter_errored(requests)
```